[waypoint] Support AuthorizationPolicy

PetrMc opened the pull request:
# Description

this is based on the issue https://github.com/kgateway-dev/kgateway/issues/10807

## API changes

No changes.

## Code changes

added Authorization Policy informer and logic to convert the policies to Envoy filter.

## CI changes

no changes

## Docs changes

Updated example readme for Waypoint with some minor adjustments.
TODO - Document Authorization policies example.

# Context

As waypoint is part of Istio setup - it needs to apply Authz rulles to the waypoint traffic based on Istio definitions.

## Interesting decisions

This first impletemntation that has plenty of room for improvements.

## Testing steps

Running binary in kind container on my machine.

## Notes for reviewers

Please understand it's my first contribution to the code base and obviously it's far from ideal.

# Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works

lgadban commented on internal/kgateway/extensions2/plugins/waypoint/rbac.go:
```diff
@@ -0,0 +1,180 @@
+	RootNamespace = "istio-system"
+)
+
+// BuildRBACForService gives three lists of filters:
```
Is this correct? I only see two lists here

lgadban commented on internal/kgateway/extensions2/plugins/waypoint/waypointquery/waypoint_query.go:
```diff
@@ -45,11 +52,24 @@ func NewQueries(
 	gwQueries query.GatewayQueries,
 ) WaypointQueries {
 	waypointedServices, servicesByWaypoint := waypointAttachmentIndex(commonCols)
+	// authzInformer is
```
I think this is unintentional but can we either remove this or expand the comment?

lgadban commented on internal/kgateway/extensions2/plugins/waypoint/waypointquery/waypoint_query.go:
```diff
@@ -37,6 +41,9 @@ type WaypointQueries interface {
 	// GetHTTPRoutesForService fetches HTTPRoutes that have the given Service in parentRefs.
 	GetHTTPRoutesForService(kctx krt.HandlerContext, ctx context.Context, svc *Service) []query.RouteInfo
 
+	// GetAuthorizationPolicies gets all AuthorizationPolicy resources in the targetNamespace and rootNamespace.
```
I think this may be inaccurate, we're not getting all authzPols in these namespaces but a filtered list of pols in the namespaces that also target services in the targetNamespace?

PetrMc replied:
good catch! Thanks!

lgadban commented on internal/kgateway/extensions2/plugins/waypoint/rbac_filters.go:
very small nit: is it really worth having a separate file for just these 2 functions?

PetrMc replied:
Merged to rbac.go - Thanks!

lgadban commented on internal/kgateway/extensions2/plugins/waypoint/rbac_filters.go:
```diff
@@ -0,0 +1,39 @@
+
+// applyHTTPRBACFilters applies RBAC filters to an HTTP filter chain
+func applyHTTPRBACFilters(httpChain *ir.HttpFilterChainIR, httpRBAC []*ir.CustomEnvoyFilter, svc waypointquery.Service) {
+	// Apply RBAC filters regardless of the presence of proxy_protocol_authority
```
this comment might deserve some expansion for readers not as close to the waypoint/sandwich internals

stevenctl replied:
I think we can elide this comment. The whole `proxy_protocol_authority` filter is a detail we can consider orthogonal to building the authz policy. Maybe a comment near where we call these `applyXXXRBACFilters` funcs to say "these will read the source identity because of the proxy_protocol authority filter (see sandwich plugin)"

PetrMc replied:
thanks there is already comment where the function is being called.

lgadban commented on internal/kgateway/extensions2/plugins/waypoint/rbac.go:
```diff
@@ -0,0 +1,180 @@
+	gatewayName, gatewayNamespace string,
+	rootNamespace string,
+	svc *waypointquery.Service,
+) *builder.Builder {
```
I'm assuming we do the same on the GG side?
cc @stevenctl

stevenctl replied:
yes

PetrMc replied:
no follow up needed.

stevenctl commented on internal/kgateway/extensions2/plugins/waypoint/rbac_filters.go:
```diff
@@ -0,0 +1,39 @@
+	if len(httpRBAC) > 0 {
+		// Initialize CustomHTTPFilters if it's nil
+		if httpChain.CustomHTTPFilters == nil {
+			httpChain.CustomHTTPFilters = []ir.CustomEnvoyFilter{}
```
this init doesn't add anything vs just appending; you could do `make([]type, 0, capacity)` but I think just appending is fine.

PetrMc replied:
good point - Thank you! it sounds like it's ok with you to keep it as is for now but will try to use `make()` more often in future

stevenctl commented on internal/kgateway/extensions2/plugins/waypoint/rbac.go:
```diff
@@ -0,0 +1,180 @@
+	return customFiltersHelper(stage, predicate, f.GetName(), config)
+}
+
+func CustomHTTPFilters(
```
We may already have some funcs like this in the `ir` package. If not, these should probably just live there.

PetrMc replied:
i got a bit lost there but probably you're right and it can be some duplications.

stevenctl replied:
Looks like it's not duplicated, but the `ir` package would be a good home for these QoL funcs.

stevenctl commented on internal/kgateway/extensions2/plugins/waypoint/waypointquery/waypoint_authz.go:
```diff
@@ -0,0 +1,35 @@
+	"istio.io/istio/pkg/kube/krt"
+)
+
+func (w *waypointQueries) GetAuthorizationPolicies(kctx krt.HandlerContext, ctx context.Context, targetNamespace, rootNamespace string) []*istiosecurity.AuthorizationPolicy {
```
We don't need to do this now, but there is room for optimization here. Right now, this means the entire Gateway IR will be recomputed when any authz policies in these namespaces change. Instead we can:

1. Make this func accept the Service instead of the targetNamespace
2. Build an index of AuthzPolicies that actually target our Service
3. Make another func that accepts the Gateway
4. Build a similar index (could be a single index that includes `groupkind`)

This would look something like the RouteIndex 

https://github.com/kgateway-dev/kgateway/blob/2eebf1c68a5fa65143fc92376a25c8c179fee303/internal/kgateway/krtcollections/policy.go#L474

https://github.com/kgateway-dev/kgateway/blob/2eebf1c68a5fa65143fc92376a25c8c179fee303/internal/kgateway/krtcollections/policy.go#L564

lgadban replied:
I created issue just so we don't lose sight of it: https://github.com/kgateway-dev/kgateway/issues/10934

great call!

stevenctl commented on internal/kgateway/extensions2/plugins/waypoint/testdata/input/authz-ns.yaml:
```diff
@@ -0,0 +1,88 @@
+  selector:
+    app: helloworld    
+---
+apiVersion: security.istio.io/v1
```
nit: add comments about the expected output here in the input

stevenctl commented on internal/kgateway/extensions2/plugins/waypoint/testdata/input/authz-ns.yaml:
```diff
@@ -0,0 +1,88 @@
+    - operation:
+        methods:
+        - GET
+  targetRefs:
```
this test says we're applying authz policy to the Namespace but this is just doing it to 2 services. 

should this target the waypoint and maybe be `authz-gw`?

stevenctl replied:
I know we have a "global authz" feature in istio, but that targets the `GatewayClass` but unfortunately that hardcodes the istio class name; we'll need to adjust istio code or fork the `ListAuthorizationPolicies` part to support this (we might want to do that anyway as part of optimizing within krt)

PetrMc replied:
sorry it was my misunderstanding of test definitions. I read it ns - Ambient enabled (labeled) at the namespace level and service enabled/labeled per service. I'll keep one test for now then. As another one will keep it open as part of the enchancment effort after this PR.

stevenctl commented on internal/kgateway/extensions2/plugins/waypoint/rbac.go:
```diff
@@ -0,0 +1,180 @@
+
+const (
+	// TODO RootNamespace should be equiv to what istio sees as root ns
+	RootNamespace = "istio-system"
```
I think maybe we can read this from an env with `istio-system` as a default? Not sure if there are other places that currently care about this but there will be in the future.

stevenctl replied:
Or from `settings`

PetrMc replied:
changed it to a varable and added to the settings, defaulting to istio-system.

PetrMc replied:
https://github.com/kgateway-dev/kgateway/blob/2914f7ad58922c98d9b45d4a9b8ce4fb90eabf4b/internal/kgateway/extensions2/settings/settings.go#L21

stevenctl commented on internal/kgateway/extensions2/plugins/waypoint/rbac.go:
```diff
@@ -0,0 +1,180 @@
+			{
+				Name:      svc.GetName(),
+				Namespace: svc.GetNamespace(),
+				Registry:  provider.Kubernetes,
```
This should be `provider.External` if the `svc` is a ServiceEntry. Maybe add a `(s *Service) Provider()` method  in `service_model.go`.

PetrMc replied:
added the function https://github.com/PetrMc/kgateway/blob/2914f7ad58922c98d9b45d4a9b8ce4fb90eabf4b/internal/kgateway/extensions2/plugins/waypoint/waypointquery/service_model.go#L142

PetrMc commented:
@lgadban - i think failing tests are flakiness. But i don't have permissions to rerun. Please let me know.

Added unit tests per our slack conversation and addressed most of the comments - there are 1-2 minor ones that seem to be ok not to act on and couple major once that we can start working on when the release is out.
@stevenctl and @lgadban - please let me know if there are still some blockers to merge this PR.

lgadban commented:
> i think failing tests are flakiness. But i don't have permissions to rerun. Please let me know.

Will kick all failed tests but the conformance failure is a bit concerning as we haven't seen flakiness on this previously

PetrMc commented:
it's me after all - adjusting kind shell script for my machine and then committing it. Sorry!

stevenctl commented on internal/kgateway/extensions2/plugins/waypoint/rbac.go:
```diff
@@ -0,0 +1,219 @@
+		return nil
+	}
+
+	trustBundle := trustdomain.NewBundle("cluster.local", nil)
```
We should probably put this `cluster.local` in a constant named defaultTrustDomain with a TODO on setting up a way to control trustDomain and trustDomainAliases.

PetrMc commented:
@lgadban can you please have a look and if all good approve. cc @stevenctl

lgadban commented on test/kubernetes/e2e/features/waypoint/testdata/authz-l7.yaml:
```diff
@@ -0,0 +1,23 @@
+apiVersion: security.istio.io/v1
+kind: AuthorizationPolicy
+metadata:
+  name: authz-l7-demy-notcurl-svc-b
```
`s/demy/deny` ?

lgadban reviewed (approved):
Excellent work on this
